include/cetl/span.h

```diff
@@ -1 +1,772 @@
+        static_assert((Count == cetl::dynamic_extent) || (Count <= extent - Offset),
+                      "subspan Count argument is ill-formed");
+        return span < element_type,
+               Count != cetl::dynamic_extent
+                   ? Count
+                   : extent - Offset > {data_ + Offset, Count == cetl::dynamic_extent ? size() - Offset : Count};
+    }
+
+    /// Create a new span `Offset` elements from the start of the current span and for either `Count` elements or,
+    /// if Count is `cetl::dynamic_extent`, for the remaining size of the span (i.e. `size() - Offset`).
+    /// The behavior of this method is undefined where `Count` is greater-than the size of this span or if
+    /// `Count` is not cetl::dynamic_extent but `Count` is greater-than `size() - Offset`.
+    /// @param Offset  Number of elements from the start of the current span for the subspan.
+    /// @param Count   The number of elements from the Offset to include in the subspan. If this value is
+    ///                `cetl::dynamic_extent` then the Count is `size() - Offset`.
+    /// @return A new span with a dynamic extent.
+    /// @see std::span::subspan()
+    constexpr span<element_type, cetl::dynamic_extent> subspan(size_type Offset,
+                                                               size_type Count = cetl::dynamic_extent) const
+    {
+        CETL_DEBUG_ASSERT(Offset <= extent, "subspan Offsets > size() are ill-formed.");
+        if (Count == cetl::dynamic_extent)
+        {
+            return {data_ + Offset, extent - Offset};
+        }
+        else
+        {
+            CETL_DEBUG_ASSERT(Count <= (extent - Offset), "subspan Count argument is ill-formed");
+            return {data_ + Offset, Count};
+        }
+    }
+
+    ///@}
+private:
+    pointer data_;
+};
+
+/// Specialization of span where the extent is dynamic.
+template <typename T>
+class span<T, cetl::dynamic_extent>
+{
+public:
+    // (Groupings per https://en.cppreference.com/w/cpp/container/span)
+    // +----------------------------------------------------------------------+
+    // | Member Constants
+    // +----------------------------------------------------------------------+
+    /// The value of Extent for this template instantiation.
+    static constexpr std::size_t extent = cetl::dynamic_extent;
+
+    // +----------------------------------------------------------------------+
+    // | Member Types
+    // +----------------------------------------------------------------------+
+    using element_type     = T;                                 //!< The element type over which the span operates
+    using value_type       = typename std::remove_cv<T>::type;  //!< The non-cv type of elements
+    using pointer          = T*;                                //!< The element pointer type
+    using const_pointer    = const T*;                          //!< A constant pointer to the element type
```

Comment:
```suggestion
    using pointer_const    = T const*;                          //!< A pointer to the constant element type
```

Reply:
The pointer is not const, the thing it points to is.

Reply:
This is per ISO/IEC. Sorry
<img width="367" alt="image" src="https://user-images.githubusercontent.com/19180861/224231021-505a29f2-42d9-4281-830d-92520ddb079a.png">